pkg/descheduler/strategies/topologyspreadyconstraint.go

```diff
@@ -0,0 +1,150 @@
+	if !strategy.Enabled {
+		return
+	}
+
+	fmt.Printf("Found following parameters for TopologySpreadConstraint %v\n", strategy)
+	for _, topoConstraints := range strategy.Params.NamespacedTopologySpreadConstraints {
+		glog.Infof("Found Topo Constraints %v\n", topoConstraints)
+		evictPodsViolatingSpreadConstraints(ds.Client, policyGroupVersion, nodes, ds.DryRun, nodePodCount, topoConstraints)
+	}
+
+}
+
+// AntiAffinityTerm's topology key value used in predicate metadata
+type topologyPair struct {
+	key   string
+	value string
+}
+
+type podSet map[*v1.Pod]struct{}
+
+type topologyPairSet map[topologyPair]struct{}
+
+// finnd all nodes
+// find all pods on the nodes
+// find the pods that match the namespace
+// find the constraints topokey , use this topokey to finds the node topokey value
+// add the pod with key as the combination of topokey and its value
+// find the min value of pods with any key
+// iterate through all keys and diff currentPods -minPods <=maxSkew
+// if diff > maxSkew, select a pod in the current bucket for eviction
+//
+func evictPodsViolatingSpreadConstraints(client clientset.Interface, policyGroupVersion string, nodes []*v1.Node, dryRun bool, nodePodCount nodePodEvictedCount,
+	constraint api.NamespacedTopologySpreadConstraint) {
+
+	if len(constraint.TopologySpreadConstraints) != 1 {
+		glog.V(1).Infof("We currently only support 1 topology spread constraint per namespace")
+		return
+	}
+	topologyPairToPods := make(map[topologyPair]podSet)
+	for _, node := range nodes {
+		glog.V(1).Infof("Processing node: %#v\n", node.Name)
+		pods, err := podutil.ListEvictablePodsOnNode(client, node, false)
+		if err != nil {
+		}
+		for _, pod := range pods {
+			if pod.Namespace != constraint.Namespace {
+				continue
+			}
+			// does this pod labels match the constraint label selector
+			// TODO: This is intentional that it only looks at the first constraint
+			selector, err := metav1.LabelSelectorAsSelector(constraint.TopologySpreadConstraints[0].LabelSelector)
+			if err != nil {
+				continue
+			}
+			if !selector.Matches(labels.Set(pod.Labels)) {
+				continue
+			}
+			// TODO: Need to determine if the topokey already present in the node or not
+			pair := topologyPair{key: constraint.TopologySpreadConstraints[0].TopologyKey, value: node.Labels[constraint.TopologySpreadConstraints[0].TopologyKey]}
+			addTopologyPair(topologyPairToPods, pair, pod)
+		}
+	}
+
+	minPodsForGivenTopo := math.MaxInt32
+	for _, v := range topologyPairToPods {
+		if len(v) < minPodsForGivenTopo {
+			minPodsForGivenTopo = len(v)
+		}
+	}
+
+	for _, v := range topologyPairToPods {
+		podsInTopo := len(v)
+		if int32(podsInTopo-minPodsForGivenTopo) >= constraint.TopologySpreadConstraints[0].MaxSkew {
+			//we need to evict maxSkew-(podsInTopo-minPodsForGivenTopo))
+			countToEvict := constraint.TopologySpreadConstraints[0].MaxSkew - int32(podsInTopo-minPodsForGivenTopo)
+			podsListToEvict := GetPodsToEvict(countToEvict, v)
```

Comment:
When multiple constraints are considered, it'd good to postpone the pickup process, so that it would be possible to pick the same Pod satisfying multiple constraints. For example, for a 2-constraint case, the calculation result decides to evict 3/2 pods. If we postpone the pickup logic, we may come up with a solution evicting 3 pods, instead of 5(3+2) pods.